crates/bevy_ui/src/keyboard_navigation.rs

```diff
@@ -0,0 +1,173 @@
+use crate::{Interaction, UiStack};
+use bevy_a11y::Focus;
+use bevy_ecs::prelude::Component;
+use bevy_ecs::query::With;
+use bevy_ecs::system::{Local, Query, Res, ResMut};
+use bevy_ecs::{change_detection::DetectChangesMut, entity::Entity};
+use bevy_input::{prelude::KeyCode, Input};
+use bevy_reflect::{FromReflect, Reflect};
+use bevy_render::view::ComputedVisibility;
+
+/// A component that represents if a UI element is focused.
+#[derive(Reflect, FromReflect, Component, Clone, Debug, Default, Eq, PartialEq)]
+pub struct Focusable {
+    focus_state: FocusState,
+}
+
+impl Focusable {
+    /// The entity is currently focused, similar to the `:focus` css pseudo-class.
+    /// To check if the focus has been achieved through keyboard navigation, see [`Focusable::is_focus_visible`].
+    pub fn is_focused(&self) -> bool {
+        matches!(self.focus_state, FocusState::Focused { .. })
+    }
+
+    /// Focus has been reached through keyboard navigation and so a focus style should be displayed.
+    /// This is similar to the `:focus-visible` pseudo-class in css.
+    pub fn is_focus_visible(&self) -> bool {
+        matches!(self.focus_state, FocusState::Focused { visible: true })
+    }
+}
+
+#[derive(Reflect, FromReflect, Component, Clone, Debug, Default, Eq, PartialEq)]
+enum FocusState {
+    /// Entity is not focused
+    #[default]
+    None,
+    /// Entity is focused
+    Focused {
+        /// Focus has been reached through keyboard navigation and so a focus style should be displayed.
+        /// This is similar to the `:focus-visible` pseudo-class in css.
+        visible: bool,
```

Comment:
Why make this a field of `FocusState::Focused` instead of a field of `Focusable`?

Reply:
It doesn't really make sense to have an entity that is not focused, but at the same time has the focus as visible. The visibility of the focus is only applicable when the entity is actually focused.

I'm happy to change this if you prefer, but I don't see any compelling reasons why having this as a field of `Focusable` would be advantageous.

Reply:
I approved the PR. I'll leave it up to you.

Personally I try to avoid boolean fields generally. It's a bad pattern. It especially makes little sense as enum field. It can be replaced by:

```rust
enum FocusState {
  FocusedVisible,
  FocusedInvisible,
  None,
```

Also I don't think `visible` is the appropriate term here. Semantically I think it means "last interaction was through a keyboard navigation interaction". Whether it's visible or not is up to the end user. Within bevy code, using "visible" is just plain misleading. I wouldn't hold the web standards as a reference for naming things, it's notoriously bad in this regard (eg: referer header).

Also I still am doubtful this is something we want for bevy. Can you give me the example of a game with tab-based keyboard navigation?

Reply:
> Can you give me the example of a game with tab-based keyboard navigation?

Games with form style interfaces (like username / password input boxes e.g. Veloren) will have at least have some kind of tab navigation.

Reply:
Also if bevy UI is used to make a level editor, then tab navigation will be necessary for that.